File: sdk/reproit-tui-py/tests/test_causal.py
import io
import json
import os
import tempfile
import urllib.request

from reproit_tui_py.causal import _redact, install_causal_urllib


def test_explicit_secret_keys_redact_without_hiding_ordinary_keys():
    safe = _redact({
        "apiKey": "raw-api", "publishable-key": "raw-pub", "private_key": "raw-private",
        "access.key": "raw-access", "signing key": "raw-signing",
        "keyboardLayout": "dvorak", "key": "ordinary",
    })
    assert safe["keyboardLayout"] == "dvorak" and safe["key"] == "ordinary"
    assert not any(raw in json.dumps(safe) for raw in ["raw-api", "raw-pub", "raw-private", "raw-access", "raw-signing"])


class FakeResponse:
    status = 200
    headers = {"Content-Type": "application/json"}

    def read(self):
        return json.dumps({"profile": {"email": "a@example.com"}, "ok": True}).encode()

    def close(self):
        pass


def test_capture_uses_side_files_and_redacts_before_persistence():
    with tempfile.TemporaryDirectory() as directory:
        network = os.path.join(directory, "network.ndjson")
        action = os.path.join(directory, "action.txt")
        capabilities = os.path.join(directory, "capabilities.json")
        open(network, "w").close()
        with open(action, "w") as handle:
            handle.write("3")
        with open(capabilities, "w") as handle:
            handle.write("{}")
        prior = urllib.request.urlopen
        urllib.request.urlopen = lambda *_a, **_k: FakeResponse()
        os.environ.update({
            "REPROIT_NETWORK_FILE": network,
            "REPROIT_ACTION_FILE": action,
            "REPROIT_CAPABILITIES_FILE": capabilities,
            "REPROIT_DEVICE": "b",
        })
        restore = install_causal_urllib()
        try:
            request = urllib.request.Request(
                "https://app.test/feed",
                data=json.dumps({"token": "raw"}).encode(),
                headers={"Authorization": "raw", "Content-Type": "application/json"},
                method="POST",
            )
            assert urllib.request.urlopen(request).status == 200
            with open(network) as handle:
                exchange = json.loads(handle.read())
            assert exchange["actor"] == "b" and exchange["actionIndex"] == 3
            assert exchange["requestBody"]["token"] == "<reproit:string:length=3>"
            assert exchange["responseBody"]["profile"]["email"] == "<reproit:string:length=13>"
            with open(capabilities) as handle:
                assert json.load(handle)["http"]["status"] == "captured"
        finally:
            restore()
            urllib.request.urlopen = prior
            for key in ["REPROIT_NETWORK_FILE", "REPROIT_ACTION_FILE", "REPROIT_CAPABILITIES_FILE", "REPROIT_DEVICE"]:
                os.environ.pop(key, None)

File: runners/test_causal.c
#define _POSIX_C_SOURCE 200809L
#define REPROIT_CAUSAL_IMPLEMENTATION
#include "reproit_causal.h"
#include <assert.h>
#include <sys/stat.h>
#include <unistd.h>

static int live(void *user, const char *method, const char *url, const char *qh, const char *qb,
                char *rh, size_t rhcap, char *rb, size_t rbcap) {
  (void)user;
  (void)method;
  (void)url;
  (void)qh;
  (void)qb;
  snprintf(rh, rhcap, "{\"content-type\":\"application/json\",\"set-cookie\":\"raw\"}");
  snprintf(rb, rbcap, "{\"ok\":true,\"email\":\"a@b.c\"}");
  return 201;
}
int main(void) {
  char dir[256], network[300], action[300], caps[300], capsule[300];
  snprintf(dir, sizeof dir, "/tmp/reproit-causal-%d", (int)getpid());
  mkdir(dir, 0700);
  snprintf(network, sizeof network, "%s/network.jsonl", dir);
  snprintf(action, sizeof action, "%s/action", dir);
  snprintf(caps, sizeof caps, "%s/caps.json", dir);
  snprintf(capsule, sizeof capsule, "%s/capsule.json", dir);
  FILE *f = fopen(action, "wb");
  fputs("1", f);
  fclose(f);
  setenv("REPROIT_NETWORK_FILE", network, 1);
  setenv("REPROIT_ACTION_FILE", action, 1);
  setenv("REPROIT_CAPABILITIES_FILE", caps, 1);
  setenv("REPROIT_DEVICE", "a", 1);
  assert(ReproIt_Causal_Enable());
  char rh[1024], rb[1024];
  assert(ReproIt_Causal_Json(
             "POST", "https://api.test/x", "{\"authorization\":\"raw\"}",
             "{\"token\":\"raw\",\"apiKey\":\"raw-api\",\"publishable-key\":\"raw-pub\",\"private_"
             "key\":\"raw-private\",\"access.key\":\"raw-access\",\"signing "
             "key\":\"raw-signing\",\"keyboardLayout\":\"dvorak\",\"key\":\"ordinary\",\"kind\":"
             "\"x\"}",
             rh, sizeof rh, rb, sizeof rb, live, NULL) == 201);
  assert(strstr(rb, "a@b.c")); /* capture redaction must not mutate the live app response */
  char *line = reproit_causal_read(network);
  assert(line && strstr(line, "<reproit:string:length=3>") && !strstr(line, "a@b.c"));
  assert(strstr(line, "keyboardLayout") && strstr(line, "dvorak") && strstr(line, "ordinary"));
  assert(!strstr(line, "raw-api") && !strstr(line, "raw-pub") && !strstr(line, "raw-private") &&
         !strstr(line, "raw-access") && !strstr(line, "raw-signing"));
  free(line);
  f = fopen(capsule, "wb");
  fputs("{\"exchanges\":[{\"id\":\"a-1-0\",\"actor\":\"a\",\"actionIndex\":1,\"ordinal\":0,"
        "\"method\":\"GET\",\"url\":\"https://api.test/"
        "config?a=1&b=2\",\"status\":200,\"responseHeaders\":{\"content-type\":\"application/"
        "json\"},\"responseBody\":{\"enabled\":true},\"required\":true}]}",
        f);
  fclose(f);
  free(reproit_causal_capsule);
  reproit_causal_capsule = reproit_causal_read(capsule);
  reproit_causal_prior_action = ~0u;
  reproit_causal_ordinal = 0;
  assert(ReproIt_Causal_Json("GET", "https://api.test/config?b=2&a=1", "{}", "null", rh, sizeof rh,
                             rb, sizeof rb, NULL, NULL) == 200);
  assert(strstr(rb, "enabled"));
  assert(ReproIt_Causal_Json("GET", "https://api.test/miss", "{}", "null", rh, sizeof rh, rb,
                             sizeof rb, NULL, NULL) == -1);
  remove(network);
  remove(action);
  remove(caps);
  remove(capsule);
  rmdir(dir);
  return 0;
}
